src/control/system/database.go

```diff
@@ -724,6 +725,103 @@ func (db *Database) FaultDomainTree() *FaultDomainTree {
 	return db.data.Members.FaultDomains.Copy()
 }
 
+// CompressedFaultDomainTree returns the tree of fault domains of joined
+// members in a compressed format.
+// Each domain is represented as a tuple: (level, ID, number of children)
+// Except for the rank, which is represented as: (rank)
+// The order of items is a breadth-first traversal of the tree.
+func (db *Database) CompressedFaultDomainTree(ranks ...uint32) ([]uint32, error) {
+	tree := db.FaultDomainTree()
+	if tree == nil {
+		return nil, errors.New("uninitialized fault domain tree")
+	}
+
+	subtree, err := getFaultDomainSubtree(tree, ranks...)
+	if err != nil {
+		return nil, err
+	}
+
+	return compressTree(subtree), nil
+}
+
+func getFaultDomainSubtree(tree *FaultDomainTree, ranks ...uint32) (*FaultDomainTree, error) {
+	if len(ranks) == 0 {
+		return tree, nil
+	}
+
+	domains := tree.Domains()
+
+	// Traverse the list of domains only once
+	treeDomains := make(map[uint32]*FaultDomain, len(ranks))
+	for _, d := range domains {
+		if r, isRank := getFaultDomainRank(d); isRank {
+			treeDomains[r] = d
+		}
+	}
+
+	rankDomains := make([]*FaultDomain, 0)
+	for _, r := range ranks {
+		d, ok := treeDomains[r]
+		if !ok {
+			return nil, fmt.Errorf("rank %d not found", r)
```

Comment:
Just wondering if more context on this error would be helpful.  I suppose someone will see the line number and can correlate it later.  But maybe there's an opportunity for giving more helpful information here.

Reply:
I can add a little more detail to the message.